wrappers/core18.go

```diff
@@ -657,6 +661,50 @@ func undoSnapdDbusActivationOnCore() error {
 	return err
 }
 
+func snapdDesktopFileNames(s *snap.Info) ([]string, error) {
+	paths, err := filepath.Glob(filepath.Join(s.MountDir(), "usr/share/applications/*.desktop"))
+	if err != nil {
+		return nil, err
+	}
+
+	fileNames := make([]string, len(paths))
+	for i, path := range paths {
+		fileNames[i] = filepath.Base(path)
+	}
+
+	return fileNames, nil
+}
+
+func writeSnapdDesktopFilesOnCore(s *snap.Info) error {
+	// Ensure /var/lib/snapd/desktop/applications exists
+	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
+		return err
+	}
+
+	desktopFileNames, err := snapdDesktopFileNames(s)
+	if err != nil {
+		return err
+	}
+
+	desktopFiles := make(map[string]osutil.FileState, len(desktopFileNames))
+	for _, fileName := range desktopFileNames {
+		desktopFiles[fileName] = &osutil.FileReference{Path: filepath.Join(s.MountDir(), "usr/share/applications", fileName)}
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDesktopFilesDir, desktopFileNames, desktopFiles)
+	return err
+}
+
+func undoSnapdDesktopFilesOnCore(s *snap.Info) error {
```

Comment:
If we have the desktop prefix as mentioned above, maybe we can reuse [`RemoveSnapDesktopFiles`](https://github.com/snapcore/snapd/blob/cd03bb167398bf86a11998511a1f2f07a366f8d6/wrappers/desktop.go#L289C6-L289C28)?